### PR TITLE
wip: Back-ported HttpProxyConnectHandler

### PR DESCRIPTION
**Pull request by spockz**
Problem

httpProxyTo is not available for the current Netty3 stack.

Solution

Back ported HttpProxyConnectHandler.

Result

ProxyTo should work for netty3 now as well.

### CODE REVIEW

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/Netty3Transporter.scala [hunk 1]
```diff
@@ -2,30 +2,32 @@ package com.twitter.finagle.netty3
...
-import com.twitter.finagle.{Stack, WriteException, CancelledConnectionException}
-import com.twitter.util.{Future, Promise, NonFatal, Stopwatch}
+import com.twitter.finagle.{CancelledConnectionException, Stack, WriteException}
+import com.twitter.util.{Future, NonFatal, Promise, Stopwatch}
 import java.net.{InetSocketAddress, SocketAddress}
 import java.nio.channels.UnresolvedAddressException
 import java.util.IdentityHashMap
 import java.util.concurrent.TimeUnit
 import java.util.logging.Level
+
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/Netty3Transporter.scala]: Not sure we need this blank line.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+ *       channel handlers that write on `channelAdded` or `channelActive`.
+ *
+ * [1]: http://www.web-cache.com/Writings/Internet-Drafts/draft-luotonen-web-proxy-tunneling-01.txt
+ * [2]: http://wiki.squid-cache.org/Features/HTTPS
+ * [3]: https://github.com/netty/netty/blob/4.1/handler-proxy/src/main/java/io/netty/handler/proxy/HttpProxyHandler.java
+ * @param host the ultimate host a remote proxy server connects to
+ * @param credentialsOption optional credentials for a proxy server
+ */
+private[netty3] class HttpProxyConnectHandler(
+                                              host: String,
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: De-indent please.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+ *
+ * This enables "Tunneling TCP-based protocols (i.e., TLS/SSL) through Web proxy servers" [1] and
+ * may be used with any TCP traffic, not only HTTP(S). See Squid documentation on this feature [2].
+ *
+ * @note We don't use Netty's implementation [3] here because it supports an opposite direction: the
+ *       destination passed to `Channel.connect` is an ultimate target and the `HttpProxyHandler`
+ *       is supposed to replace it with proxy addr (represented as a `SocketAddress`). This is the
+ *       exact approach we used for Netty 3 implementation, but we don't do that anymore because we
+ *       don't want to bypass Finagle's load balancers while resolving the proxy endpoint.
+ * @note This mixes in a [[BufferingChannelOutboundHandler]] so we can protect ourselves from
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: No, it doesn't.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+/**
+ * An internal handler that upgrades the pipeline to delay connect-promise satisfaction until the
+ * remote HTTP proxy server is ready to proxy traffic to an ultimate destination represented as
+ * `host` (i.e., HTTP proxy connect procedure is successful).
+ *
+ * This enables "Tunneling TCP-based protocols (i.e., TLS/SSL) through Web proxy servers" [1] and
+ * may be used with any TCP traffic, not only HTTP(S). See Squid documentation on this feature [2].
+ *
+ * @note We don't use Netty's implementation [3] here because it supports an opposite direction: the
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Just feel free to remove this comment.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+  }
+
+  private[this] def fail(c: Channel, t: SocketAddress => Throwable) {
+    Option(connectFuture.get) foreach { _.setFailure(t(c.getRemoteAddress)) }
+    Channels.close(c)
+  }
+
+  override def connectRequested(ctx: ChannelHandlerContext, e: ChannelStateEvent): Unit = {
+    println("entering connectRequested")
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Remove?

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+  private[this] val httpCodecKey: String = "http proxy client codec"
+
+  private[this] val connectFuture = new AtomicReference[ChannelFuture](null)
+
+  private[this] def proxyAuthorizationHeader(c: Credentials): String = {
+    val bytes = "%s:%s".format(c.username, c.password).getBytes(Charsets.Utf8)
+    "Basic " + Base64StringEncoder.encode(bytes)
+  }
+
+  private[this] def fail(ctx: ChannelHandlerContext, t: SocketAddress => Throwable): Unit = {
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Do we need both `fail`s?

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+              return
+            }
+
+            fail(f.getChannel, (_: SocketAddress) => f.getCause)
+          }
+        })
+
+        // We propagate the pipeline with a new promise thereby delaying the original connect's
+        // satisfaction.
+        // TODO: Figure out why the remoteAddress is
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Remote address what?

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        val wrappedEvent = new DownstreamChannelStateEvent(
+          de.getChannel, wrappedConnectFuture,
+          de.getState, de.getValue)
+
+
+        println(s"Connecting to ${wrappedEvent.getChannel.getRemoteAddress} with $wrappedEvent")
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Remove?

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+  private[this] def fail(c: Channel, t: SocketAddress => Throwable) {
+    Option(connectFuture.get) foreach { _.setFailure(t(c.getRemoteAddress)) }
+    Channels.close(c)
+  }
+
+  override def connectRequested(ctx: ChannelHandlerContext, e: ChannelStateEvent): Unit = {
+    println("entering connectRequested")
+    e match {
+      case de: DownstreamChannelStateEvent =>
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: I think we need an extra match to cover non-downstream-channel events.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+  }
+
+  override def writeRequested(ctx: ChannelHandlerContext, e: MessageEvent): Unit = {
+    println(s"Storing $e for now")
+    super.writeRequested(ctx, e)
+  }
+
+  override def channelConnected(ctx: ChannelHandlerContext, e: ChannelStateEvent): Unit = {
+    println("channelConnected")
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Please remove.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+          de.getState, de.getValue)
+
+
+        println(s"Connecting to ${wrappedEvent.getChannel.getRemoteAddress} with $wrappedEvent")
+        super.connectRequested(ctx, wrappedEvent)
+    }
+
+  }
+
+  override def writeRequested(ctx: ChannelHandlerContext, e: MessageEvent): Unit = {
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Let's not override it. I think we're ok assuming that nothing is writing into the pipeline on N3 until the connect future is satisfied.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+  override def channelConnected(ctx: ChannelHandlerContext, e: ChannelStateEvent): Unit = {
+    println("channelConnected")
+    if (connectFuture.get eq null) {
+      fail(ctx, (remoteAddress: SocketAddress) => new InconsistentStateException(remoteAddress))
+      return
+    }
+    // Add HTTP client codec so we can talk to an HTTP proxy.
+    ctx.getPipeline().addBefore(ctx.getName(), httpCodecKey, httpClientCodec)
+
+    val httpProxyConnectRequestFuture = Channels.future(e.getChannel, true)
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: You don't need it. Just use `connectFuture` (is what Finagle transport waiting on).

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+    // Create new connect HTTP proxy connect request.
+    val req = new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.CONNECT, host)
+    req.headers().set(HttpHeaders.Names.HOST, host).set(HttpHeaders.Names.CONNECTION, "keep-alive")
+    credentialsOption.foreach(c =>
+      req.headers().add(HttpHeaders.Names.PROXY_AUTHORIZATION, proxyAuthorizationHeader(c))
+    )
+
+    println(s"Sending Connect message downstream [$req]")
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Please, remove.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+    ctx.sendDownstream(new DownstreamMessageEvent(ctx.getChannel, httpProxyConnectRequestFuture, req, ctx.getChannel.getRemoteAddress))
+    httpProxyConnectRequestFuture.addListener(new ChannelFutureListener {
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Don't need this.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+          HttpProxyConnectHandler.super.channelConnected(ctx, e)
+        } else {
+          Channels.close(ctx.getChannel)
+        }
+    })
+
+    // proxy cancellations again.
+    connectFuture.get.addListener(new ChannelFutureListener {
+      override def operationComplete(f: ChannelFuture): Unit =
+        if (f.isCancelled) {
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: You need to handle the success case here.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+          fail(ctx.getChannel, (remoteAddress : SocketAddress) =>new ChannelClosedException(remoteAddress))
+        }
+    })
+
+
+    //perhaps read is necessary
+  }
+
+  override def messageReceived(ctx: ChannelHandlerContext, messageEvent: MessageEvent): Unit =     {
+      println(s"Received Message: $messageEvent")
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Please, remove.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    // proxy cancellations again.
+    connectFuture.get.addListener(new ChannelFutureListener {
+      override def operationComplete(f: ChannelFuture): Unit =
+        if (f.isCancelled) {
+          fail(ctx.getChannel, (remoteAddress : SocketAddress) =>new ChannelClosedException(remoteAddress))
+        }
+    })
+
+
+    //perhaps read is necessary
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: It's it's fine - N3 is "auto-read" by default.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+  }
+
+  override def messageReceived(ctx: ChannelHandlerContext, messageEvent: MessageEvent): Unit =     {
+      println(s"Received Message: $messageEvent")
+      messageEvent.getMessage match {
+        case rep: DefaultHttpResponse =>
+          println(s"Received response: ${rep.toString}")
```
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: Please, remove.

File: finagle-http/src/test/scala/com/twitter/finagle/http/netty/ProxyTest.scala [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+import com.twitter.finagle.Http
+import com.twitter.finagle.http.RequestBuilder
+import com.twitter.finagle.util.DefaultTimer
+import com.twitter.io.Buf
+import com.twitter.util.{Await, Duration, Return, Throw}
+
+/**
+ * Created by alessandro on 13/07/16.
+ */
+object ProxyTest extends App {
```
Comment (vkostyukov) [inline finagle-http/src/test/scala/com/twitter/finagle/http/netty/ProxyTest.scala]: Can we backport the N4 test as well?

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala (hunk 7 above)
Comment (spockz) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: I recall that it was `null` in some cases.

File: finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala (hunk 15 above)
Comment (vkostyukov) [inline finagle-core/src/main/scala/com/twitter/finagle/netty3/proxy/HttpProxyConnectHandler.scala]: I think if you handle the success here you'd make it work with TLS/SSL.

**Pull request closed without merging**